### PR TITLE
add nft module

### LINKED ISSUE #91: NFT Module
```rust=
validate_mint_request(anchor_id, proofs, minter_address):
  roothash, timestamp = get_anchor()
  
  res = proofValidator(proofs) //github.com/centrifuge/privacy-enabled-erc721/blob/develop/src/merkle.sol
  assert(res == true)
  
  var data byte32[]
  data.append(minter_address)

  for i in len(proofs):
    data = append(data, proofs[len(proofs[i]-1)]) //get leaf proof

  // data.append(timestamp)
 
  emit(anchor_id, sha3(data))
```

- Input: 
    - proof list
    - ethereum address that will call _mint_ on Ethereum contract

- Call proof verification above (https://github.com/centrifuge/centrifuge-chain/pull/92/files)
- If success:
    - hash `minter_address` + all leaves into a b32 value ideally with ethereum compatible sha3 (keccak)
    - emit event with resultant hash
- If failure:
    - automatically creates fail event (nothing to do here)

### PR DESCRIPTION
**Pull request by vedhavyas**
Closes #91

### CODE REVIEW

File: runtime/src/nfts.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+            ensure_signed(origin)?;
+
+            // get the anchor data from anchor ID
+            let anchor_data = <anchor::Module<T>>::get_anchor_by_id(anchor_id).ok_or("Anchor doesn't exist")?;
+
+            // validate proofs
+            ensure!(Self::validate_proofs(anchor_data.get_doc_root(), &pfs), "Invalid proofs");
+
+            // get the bundled hash
+            let bh = Self::get_bundled_hash(pfs, deposit_address);
```
Comment (vimukthi-git) [inline runtime/src/nfts.rs]:
```suggestion
            let bundled_hash = Self::get_bundled_hash(pfs, deposit_address);
```

File: runtime/src/proofs.rs [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
-#[derive(Debug)]
+#[derive(Encode, Decode, Default, Clone, PartialEq)]
+#[cfg_attr(feature = "std", derive(Debug))]
 pub struct Proof {
     hash: [u8; 32],
     sorted_hashes: Vec<[u8; 32]>,
 }
 
+impl Proof {
+    pub fn new(hash: [u8; 32], sorted_hashes: Vec<[u8; 32]>) -> Self {
```
Comment (vimukthi-git) [inline runtime/src/proofs.rs]: I think this constructor is not necessary

File: runtime/src/nfts.rs [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+}
+
+impl<T: Trait> Module<T> {
+    fn validate_proofs(doc_root: T::Hash, pfs: &Vec<proofs::Proof>) -> bool {
+        let mut dr: [u8; 32] = Default::default();
+        dr.clone_from_slice(doc_root.as_ref());
+        proofs::validate_proofs(dr, pfs)
+    }
+
+    fn get_bundled_hash(pfs: Vec<proofs::Proof>, deposit_address: [u8; 20]) -> T::Hash {
```
Comment (vimukthi-git) [inline runtime/src/nfts.rs]: I find this name a bit confusing, what does it exactly do?

File: runtime/src/nfts.rs [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    pub enum Event<T> where <T as system::Trait>::Hash {
+        DepositAsset(Hash),
+    }
+);
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin  {
+        fn deposit_event() = default;
+
+        fn validate_mint(origin, anchor_id: T::Hash, deposit_address: [u8; 20], pfs: Vec<proofs::Proof>) -> Result {
```
Comment (vimukthi-git) [inline runtime/src/nfts.rs]: rust doc please

File: runtime/src/nfts.rs (hunk 3 above)
Comment (vedhavyas) [reply runtime/src/nfts.rs]: explained in the comments

File: runtime/src/proofs.rs (hunk 2 above)
Comment (vedhavyas) [reply runtime/src/proofs.rs]: I would like to keep it for easier inits for proofs.

**Pull request merged**